core/types/timed_cache.go

```diff
@@ -0,0 +1,177 @@
+		if v.expired() {
+			tc.cache.Remove(key)
+			return nil, false
+		} else {
+			return v.value, true
+		}
+	} else {
+		return nil, false
+	}
+}
+
+// Contains checks if a key is in the cache, without updating the
+// recent-ness or deleting it for being stale.
+func (tc *TimedCache) Contains(key interface{}) bool {
+	_, ok := tc.Peek(key)
+	return ok
+}
+
+// Peek returns the key value (or undefined if not found) without updating
+// the "recently used"-ness or ttl of the key.
+func (tc *TimedCache) Peek(key interface{}) (value interface{}, ok bool) {
+	val, ok := tc.cache.Peek(key)
+	if ok {
+		v := val.(timedEntry)
+		if v.expired() {
+			tc.cache.Remove(key)
+			return nil, false
+		} else {
+			return v.value, ok
+		}
+	} else {
+		return nil, false
+	}
+}
+
+// ContainsOrAdd checks if a key is in the cache without updating the
+// recent-ness, ttl, or deleting it for being stale, and if not, adds the value.
+// Returns whether found and whether an eviction occurred.
+func (tc *TimedCache) ContainsOrAdd(key, value interface{}) (ok, evicted bool) {
+	// First remove expired entries, so that LRU cache doesn't evict more than
+	// necessary, if there is not enough room to add this entry.
+	tc.removeExpired()
+	// Calculate expiration time
+	expiresAt := time.Now().Unix() + tc.Ttl()
+	// Wrap the entry and add it to the cache
+	return tc.cache.ContainsOrAdd(key, timedEntry{expiresAt: expiresAt, value: value})
+}
+
+// PeekOrAdd checks if a key is in the cache without updating the
+// recent-ness, ttl, or deleting it for being stale, and if not, adds the value.
+// Returns whether found and whether an eviction occurred.
+func (tc *TimedCache) PeekOrAdd(key, value interface{}) (previous interface{}, ok, evicted bool) {
+	// First remove expired entries, so that LRU cache doesn't evict more than
+	// necessary, if there is not enough room to add this entry.
+	tc.removeExpired()
+	// Calculate expiration time
+	expiresAt := time.Now().Unix() + tc.Ttl()
+	// Wrap the entry and add it to the cache
+	return tc.cache.PeekOrAdd(key, timedEntry{expiresAt: expiresAt, value: value})
+}
+
+// Remove removes the provided key from the cache.
+func (tc *TimedCache) Remove(key interface{}) (present bool) {
+	return tc.cache.Remove(key)
+}
+
+// Resize changes the cache size.
+func (tc *TimedCache) Resize(size int) (evicted int) {
+	return tc.cache.Resize(size)
+}
+
+// RemoveOldest removes the oldest item from the cache.
+func (tc *TimedCache) RemoveOldest() (key, value interface{}, ok bool) {
+	k, v, ok := tc.cache.RemoveOldest()
+	if ok {
+		v = v.(timedEntry).value
+	}
+	return k, v, ok
+}
+
+// GetOldest returns the oldest entry
+func (tc *TimedCache) GetOldest() (key, value interface{}, ok bool) {
+	k, v, ok := tc.cache.GetOldest()
+	if ok {
+		v = v.(timedEntry).value
+	}
+	return k, v, ok
+}
+
+// Keys returns a slice of the keys in the cache, from oldest to newest.
+func (tc *TimedCache) Keys() []interface{} {
+	return tc.cache.Keys()
```

Comment:
Can keys ever contain items that have expired?

Reply:
Good catch, I should remove expired before returning this result